SharedUtils/src/main/java/com/braintreepayments/api/BraintreeSharedPreferences.java

```diff
@@ -39,84 +50,105 @@ static SharedPreferences getSharedPreferences(Context context) {
                     EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                     EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
             );
-        } catch (Exception ignored) {
-            // In the rare case that we are unable to obtain a reference to encrypted shared
-            // preferences, all preference update and retrieval logic will no-op
-            return null;
+        } catch (Exception e) {
+            throw new BraintreeSharedPreferencesException(SHARED_PREFS_ERROR_MESSAGE, e);
         }
     }
 
-    String getString(Context context, String key, String fallback)  {
-        SharedPreferences sharedPreferences = getSharedPreferences(context);
-        if (sharedPreferences != null) {
-            return sharedPreferences.getString(key, fallback);
+    BraintreeSharedPreferences(SharedPreferences sharedPreferences) {
+        this.sharedPreferences = sharedPreferences;
+        this.createException = null;
+    }
+
+    BraintreeSharedPreferences(BraintreeSharedPreferencesException createException) {
+        this.sharedPreferences = null;
+        this.createException = createException;
+    }
```

Comment:
Why can we throw the exception directly on `getInstance`? Why do we store it and just move the error further? wouldnt be best to directly "inform" that shared preferences were not created?

Reply:
My guess is that there are some instances where we are required to have shared prefs like configuration and maybe the Venmo flow (my recollection could be off here). And in those instances we want to surface the error, but in other places we may not care about shared prefs as much so can just no-op when there is an error instead of surface the error. Not sure if this tracks because my android/shared prefs knowledge is rocky at best.